django_toolkit/fallbacks/circuit_breaker/__init__.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class CircuitBreaker:
+
+    def __init__(
+        self,
+        rule,
+        cache,
+        failure_exception,
+        failure_timeout=None,
+        circuit_timeout=None,
+        catch_exceptions=None,
+    ):
+        self.rule = rule
+        self.cache = cache
+        self.failure_timeout = failure_timeout
+        self.circuit_timeout = circuit_timeout
+        self.circuit_cache_key = 'circuit_{}'.format(rule.failure_cache_key)
+        self.failure_exception = failure_exception
+        self.catch_exceptions = catch_exceptions or (Exception,)
+        self.failure_cache_key = self.rule.failure_cache_key
+        self.request_cache_key = getattr(self.rule, 'request_cache_key', '')
```

Comment:
I think the `CircuitBreaker` should not have to handle the lack of `request_cache_key` attribute. Everything that's available to do in the subclass must be available in the superclass, if you are depending to the interface (It is the [Liskov Substitution Principle](https://stackoverflow.com/questions/56860/what-is-an-example-of-the-liskov-substitution-principle)). It should be a required parameter in `Rule.__init__` and set to `None` in `MaxFailuresRule.__init__`.

Maybe, you could even take advantage of that to have concrete implementations of `should_increase_failure_count` and `should_increase_request_count` in `Rule`. It could only check if `failure_cache_key ` and `request_cache_key` are not `None`.